components/formats-gpl/src/loci/formats/in/HarmonyReader.java

```diff
@@ -0,0 +1,783 @@
+
+  /* @see loci.formats.IFormatReader#getUsedFiles(boolean) */
+  @Override
+  public String[] getUsedFiles(boolean noPixels) {
+    FormatTools.assertId(currentId, true, 1);
+
+    ArrayList<String> files = new ArrayList<String>();
+    files.addAll(metadataFiles);
+    if (!noPixels) {
+      for (Plane[] well : planes) {
+        for (Plane p : well) {
+          files.add(p.filename);
+        }
+      }
+    }
+    return files.toArray(new String[files.size()]);
+  }
+
+  /* @see loci.formats.IFormatReader#getSeriesUsedFiles(boolean) */
+  @Override
+  public String[] getSeriesUsedFiles(boolean noPixels) {
+    FormatTools.assertId(currentId, true, 1);
+
+    ArrayList<String> files = new ArrayList<String>();
+    files.addAll(metadataFiles);
+    if (!noPixels) {
+      for (Plane p : planes[getSeries()]) {
+        files.add(p.filename);
+      }
+    }
+
+    return files.toArray(new String[files.size()]);
+  }
+
+  /* @see loci.formats.IFormatReader#close(boolean) */
+  @Override
+  public void close(boolean fileOnly) throws IOException {
+    super.close(fileOnly);
+    if (!fileOnly) {
+      if (reader != null) {
+        reader.close();
+      }
+      reader = null;
+      planes = null;
+      metadataFiles.clear();
+    }
+  }
+
+  /**
+   * @see loci.formats.IFormatReader#openBytes(int, byte[], int, int, int, int)
+   */
+  @Override
+  public byte[] openBytes(int no, byte[] buf, int x, int y, int w, int h)
+    throws FormatException, IOException
+  {
+    FormatTools.checkPlaneParameters(this, no, buf.length, x, y, w, h);
+
+    if (getSeries() < planes.length && no < planes[getSeries()].length) {
+      Plane p = planes[getSeries()][no];
+
+      if (new Location(p.filename).exists()) {
+        if (reader == null) {
+          reader = new MinimalTiffReader();
+        }
+        reader.setId(p.filename);
+        reader.openBytes(0, buf, x, y, w, h);
+        reader.close();
+      }
+    }
+
+    return buf;
+  }
+
+  // -- Internal FormatReader API methods --
+
+  /* @see loci.formats.FormatReader#initFile(String) */
+  @Override
+  protected void initFile(String id) throws FormatException, IOException {
+    // make sure that we have the XML file and not a TIFF file
+
+    if (!checkSuffix(id, "xml")) {
+      Location parent = new Location(id).getAbsoluteFile().getParentFile();
+      Location xml = new Location(parent, XML_FILES[0]);
+      if (!xml.exists()) {
+        throw new FormatException("Could not find XML file " +
+          xml.getAbsolutePath());
+      }
+      initFile(xml.getAbsolutePath());
+      return;
+    }
+    else {
+      super.initFile(id);
+    }
+
+    // assemble list of other metadata/analysis results files
+
+    Location parent = new Location(currentId).getAbsoluteFile().getParentFile().getParentFile();
+    String[] list = parent.list();
+    for (String f : list) {
+      Location path = new Location(parent, f);
+      if (path.isDirectory()) {
+        String[] results = path.list();
```

Comment:
naming this results might be confusing, as far as I understand the code, this would also capture assaylayout folders and potentially others that are parallel to the Images folder of the Harmony dataset. Maybe name this companionFolders?